### DISCUSSION TITLE
Update CUDA repository GPG keys on Ubuntu 18.04

### DISCUSSION THREAD

**Discussion by patrick-5546** [adamrehn/ue4-docker#294] [Q&A]
Current the following command does not work: `ue4-docker build 4.19.2 --cuda=10.0 --no-full`. This is because the [CUDA repository GPG keys on Ubuntu 18.04 have been rotated](https://forums.developer.nvidia.com/t/gpg-error-http-developer-download-nvidia-com-compute-cuda-repos-ubuntu1804-x86-64/212904). I have found a [workaround](https://github.com/microsoft/AirSim/issues/4564#issuecomment-1213595898) for a dockerfile that uses the same base image, but am unsure which dockerfiles need to be updated in this repository; would I just have to add those 2 lines after line 2 of https://github.com/adamrehn/ue4-docker/blob/master/ue4docker/dockerfiles/ue4-build-prerequisites/linux/Dockerfile?

**Comment by TBBle** [comment]
That workaround appears to just disable the CUDA repositories. If we don't need anything from them, that's fine (although then why are we using an nvidia base image at all?).

Otherwise, we might want to instead use [this process](https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771) which updates the signing keys instead.

Whichever way we go, I'd put the workaround immediately before [the first `RUN apt-get update` step in the ue4-build-prerequisites Dockerfile](https://github.com/adamrehn/ue4-docker/blob/fa2effd5f1466bd73e3b915a48e1109a00293006/ue4docker/dockerfiles/ue4-build-prerequisites/linux/Dockerfile#L24-L25) so it's clear why it's there, and what it relates to.

**Reply by patrick-5546** [reply]
I found [this](https://github.com/nytimes/rd-blender-docker/issues/41#issuecomment-1114259869) resolves the issue for Ubuntu 18.04, but am not sure how to generalize it for all Ubuntu versions and architectures that there are CUDA repos for. But technically this workaround is only needed for Ubuntu 18.04 and not the newer versions.

**Reply by TBBle** [reply]
It'd take some shell-scripting magic to populate the variables in
```
https://developer.download.nvidia.com/compute/cuda/repos/$distro/$arch/cuda-keyring_1.0-1_all.deb
```
but we since that's an `all` package, each `$arch` directory should have an identical copy of the package anyway. I suspect the package is actually going to be the same across all `$distro` values, but haven't checked that.

And if it is not, [the underlying key file](https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771#alternate-method-manually-install-new-signing-key-4) is almost certainly release-agnostic, and so we just grab it, confirm every `$distro` version is the same, commit it this repo, and `RUN apt-key` as described there.

**Reply by patrick-5546** [reply]
I have confirmed that across releases, the cuda-keyring file differs but the underlying key file is the same

**Reply by patrick-5546** [reply]
I have confirmed that the following works on cudagl images built on ubuntu 18.04 and 20.04:

```
RUN [ ! -e /etc/apt/sources.list.d/cuda.list ] \
    || rm /etc/apt/sources.list.d/cuda.list
RUN [ ! -e /etc/apt/sources.list.d/nvidia-ml.list ] \
    || rm /etc/apt/sources.list.d/nvidia-ml.list
RUN apt-key del 7fa2af80 \
    && apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/3bf863cc.pub
```

`cuda.list` and `nvidia-ml.list` doesn't exist on 20.04 and beyond.

**Reply by TBBle** [reply]
If you're using the last `RUN` line, then you won't need the previous two, the point of that last line is to make the first two unnecessary, without losing access to those repos in the meantime.

**Reply by patrick-5546** [reply]
Without the first 2 `RUN` lines, `apt-get update` still errors out:

```
W: GPG error: https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F60F4B3D7FA2AF80
E: The repository 'https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release' is not signed.
```

**Reply by TBBle** [reply]
Ah yes, the machine-learning repo is [abandoned](https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771#machine-learning-repository-11), so we do still need to delete `/etc/apt/sources.list.d/nvidia-ml.list`, as it has not been resigned with the new key.

**Reply by TBBle** [reply]
I'd also suggest that after enrolling the new key and `apt get update` being done, doing an `apt-get install cuda-keyring` (with all the ceremony of the other `apt-get install` calls) to bring the package state into line with the actual state. (That might need to be `apt upgrade -y cuda-keyring` if `apt-get install` doesn't auto-upgrade the named package.

**Comment by patrick-5546** [comment]
I just found out that `nvidia:cuda` replaces `nvidia:cudagl` and its images use the updated cuda repo signing key: https://hub.docker.com/r/nvidia/cuda.

Searching through the repository for `cudagl`, the following files should be updated to use the new image:
- `docs/advanced-build-options.adoc`
- `docs/available-container-images.adoc`
- `ue4docker/infrastructure/BuildConfiguration.py`
